File: logging_middleware/logger.py
"""
Logging Middleware — Affordmed Campus Hiring Evaluation
=======================================================
Usage:
    from logger import Log

    Log("backend", "info",  "route",   "GET /notifications received")
    Log("backend", "error", "handler", "received string, expected bool")
    Log("backend", "fatal", "db",      "Critical database connection failure.")
"""

import requests
from config import (
    AUTH_ENDPOINT, LOG_ENDPOINT,
    CLIENT_ID, CLIENT_SECRET, EMAIL, NAME, ROLL_NO, ACCESS_CODE,
    VALID_STACKS, VALID_LEVELS, BACKEND_PACKAGES, FRONTEND_PACKAGES,
)

# Token cached in memory for the session
_token_cache: dict = {}


def _get_token() -> str:
    """Authenticate once and cache the Bearer token."""
    if _token_cache.get("access_token"):
        return _token_cache["access_token"]

    payload = {
        "email":        EMAIL,
        "name":         NAME,
        "rollNo":       ROLL_NO,
        "accessCode":   ACCESS_CODE,
        "clientID":     CLIENT_ID,
        "clientSecret": CLIENT_SECRET,
    }

    try:
        res = requests.post(AUTH_ENDPOINT, json=payload, timeout=10)
        res.raise_for_status()
        data = res.json()
        token = data.get("access_token") or data.get("token", "")
        if not token:
            raise RuntimeError(f"Auth response missing token. Got: {data}")
        _token_cache["access_token"] = token
        print("[logger] ✓ Auth token obtained.")
        return token
    except requests.RequestException as exc:
        raise RuntimeError(f"[logger] Auth failed: {exc}") from exc


def _validate(stack: str, level: str, package: str) -> None:
    """Raise ValueError if any field is invalid."""
    if stack not in VALID_STACKS:
        raise ValueError(f"[logger] Invalid stack '{stack}'. Allowed: {VALID_STACKS}")
    if level not in VALID_LEVELS:
        raise ValueError(f"[logger] Invalid level '{level}'. Allowed: {VALID_LEVELS}")

    allowed = BACKEND_PACKAGES if stack == "backend" else FRONTEND_PACKAGES
    if package not in allowed:
        raise ValueError(
            f"[logger] Invalid package '{package}' for stack '{stack}'. Allowed: {allowed}"
        )


def Log(stack: str, level: str, package: str, message: str) -> dict:
    """
    Send a log entry to the Affordmed evaluation server.

    Parameters
    ----------
    stack   : "backend" | "frontend"
    level   : "debug" | "info" | "warn" | "error" | "fatal"
    package : valid package name (lowercase)
    message : descriptive message about the event

    Returns
    -------
    dict — server response with logID
    """
    _validate(stack, level, package)
    token = _get_token()

    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type":  "application/json",
    }
    payload = {
        "stack":   stack,
        "level":   level,
        "package": package,
        "message": message,
    }

    try:
        res = requests.post(LOG_ENDPOINT, json=payload, headers=headers, timeout=10)
        res.raise_for_status()
        result = res.json()
        print(
            f"[logger] [{stack.upper()}][{level.upper()}][{package}] "
            f"{message}  → logID={result.get('logID', 'N/A')}"
        )
        return result
    except requests.HTTPError as exc:
        print(f"[logger] HTTP error: {exc.response.status_code} {exc.response.text}")
        raise
    except requests.RequestException as exc:
        print(f"[logger] Network error: {exc}")
        raise

File: notification_app_fe/src/middleware/logger.js
/**
 * Logging Middleware — Frontend (JavaScript)
 * Mirrors the Python logging_middleware package.
 *
 * Usage:
 *   import { Log } from './middleware/logger';
 *   await Log("frontend", "info", "component", "PriorityInbox mounted");
 */

const LOG_ENDPOINT  = "/api/logs";
const AUTH_ENDPOINT = "/api/auth";

// Credentials — pulled from .env (Vite exposes VITE_* vars)
const CREDS = {
  email:        import.meta.env.VITE_EMAIL        || "",
  name:         import.meta.env.VITE_NAME         || "",
  rollNo:       import.meta.env.VITE_ROLL_NO      || "",
  accessCode:   import.meta.env.VITE_ACCESS_CODE  || "",
  clientID:     import.meta.env.VITE_CLIENT_ID    || "",
  clientSecret: import.meta.env.VITE_CLIENT_SECRET || "",
};

const VALID_STACKS   = new Set(["backend", "frontend"]);
const VALID_LEVELS   = new Set(["debug", "info", "warn", "error", "fatal"]);
const FRONTEND_PKGS  = new Set([
  "api", "component", "hook", "page", "state", "style",
  "auth", "config", "middleware", "utils",
]);

let _token = null;

async function _getToken() {
  if (_token) return _token;

  const res = await fetch(AUTH_ENDPOINT, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(CREDS),
  });

  if (!res.ok) {
    const txt = await res.text();
    throw new Error(`[logger] Auth failed ${res.status}: ${txt}`);
  }

  const data  = await res.json();
  const token = data.access_token || data.token;
  if (!token) throw new Error("[logger] Auth response missing access_token");

  _token = token;
  console.info("[logger] ✓ Auth token obtained.");
  return _token;
}

// Exported so hooks/api can reuse the same token
export async function getAuthToken() {
  return _getToken();
}

function _validate(stack, level, pkg) {
  if (!VALID_STACKS.has(stack))
    throw new Error(`[logger] Invalid stack '${stack}'`);
  if (!VALID_LEVELS.has(level))
    throw new Error(`[logger] Invalid level '${level}'`);
  if (!FRONTEND_PKGS.has(pkg))
    throw new Error(`[logger] Invalid package '${pkg}' for frontend`);
}

/**
 * Log(stack, level, pkg, message)
 * @returns {Promise<{logID: string}>}
 */
export async function Log(stack, level, pkg, message) {
  try {
    _validate(stack, level, pkg);
    const token = await _getToken();

    const res = await fetch(LOG_ENDPOINT, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ stack, level, package: pkg, message }),
    });

    if (!res.ok) {
      console.error(`[logger] Log POST failed ${res.status}`);
      return null;
    }

    const result = await res.json();
    console.log(
      `[logger] [${stack.toUpperCase()}][${level.toUpperCase()}][${pkg}] ${message}`,
      `→ logID=${result.logID}`
    );
    return result;
  } catch (err) {
    // Never crash the app over a failed log
    console.warn("[logger] Suppressed log error:", err.message);
    return null;
  }
}